sktree/tree/_honest_tree.py

```diff
@@ -0,0 +1,507 @@
+
+        # update the number of classes, unsplit
+        if y.ndim == 1:
+            # reshape is necessary to preserve the data contiguity against vs
+            # [:, np.newaxis] that does not.
+            y = np.reshape(y, (-1, 1))
+        check_classification_targets(y)
+        y = np.copy(y).astype(int)
+        # Normally called by super
+        X = self.estimator._validate_X_predict(X, True)
+        # Fit leaves using other subsample
+        honest_leaves = self.tree_.apply(X[self.honest_indices_])
+
+        self._set_leaf_nodes(honest_leaves, y)
+
+        # preserve from underlying tree
+        # https://github.com/scikit-learn/scikit-learn/blob/1.0.X/sklearn/tree/_classes.py#L202
+        self._tree_classes_ = self.classes_
+        self._tree_n_classes_ = self.n_classes_
+        self.classes_ = []
+        self.n_classes_ = []
+        self.empirical_prior_ = []
+
+        y_encoded = np.zeros(y.shape, dtype=int)
+        for k in range(self.n_outputs_):
+            classes_k, y_encoded[:, k] = np.unique(y[:, k], return_inverse=True)
+            self.classes_.append(classes_k)
+            self.n_classes_.append(classes_k.shape[0])
+            self.empirical_prior_.append(
+                np.bincount(y_encoded[:, k], minlength=classes_k.shape[0]) / y.shape[0]
+            )
+        y = y_encoded
+
+        self.n_classes_ = np.array(self.n_classes_, dtype=np.intp)
+
+        if self.n_outputs_ == 1:
+            self.n_classes_ = self.n_classes_[0]
+            self.classes_ = self.classes_[0]
+            self.empirical_prior_ = self.empirical_prior_[0]
+            y = y[:, 0]
+
+        return self
+
+    def _set_leaf_nodes(self, X, y):
+        "Traverse the already built tree with X and set leaf nodes with y"
+        self.tree_.value[:, :, :] = 0
+        for leaf_id, yval in zip(X, y[self.honest_indices_, 0]):
+            self.tree_.value[leaf_id][0, yval] += 1
+
+    def _inherit_estimator_attributes(self):
+        """Initialize necessary attributes from the provided estimator"""
+        self.classes_ = self.estimator.classes_
+        self.max_features_ = self.estimator.max_features_
+        self.n_classes_ = self.estimator.n_classes_
+        self.n_features_in_ = self.estimator.n_features_in_
+        self.n_outputs_ = self.estimator.n_outputs_
+        self.tree_ = self.estimator.tree_
+
+    def _empty_leaf_correction(self, proba, normalizer):
+        """Leaves with empty posteriors are assigned values"""
+        zero_mask = proba.sum(axis=1) == 0.0
+        if self.honest_prior == "empirical":
+            proba[zero_mask] = self.empirical_prior_
+        elif self.honest_prior == "uniform":
+            proba[zero_mask] = 1 / self.n_classes_
+        elif self.honest_prior == "ignore":
+            proba[zero_mask] = np.nan
+        else:
+            raise ValueError(f"honest_prior {self.honest_prior} not a valid input.")
+
+        return proba
+
+    def _impute_missing_classes(self, proba):
```

Comment:
Can we add a test for this function when building a tree?

Reply:
Currently the function seems to be a failsafe for trees that don't use the whole y labels, similar to what we discussed earlier about subsampling. Right now, however, none of our tree estimators seem to need this function... I would just keep it for future proof, but I don't know how we can test it now.

Reply:
I see but implicitly with the sample weight hack this situation never occurs right? 

Let's just remove it then? I don't think we'll even need in future. 

For propensity tree the only thing we need is the refitting leaves function.